Feature/issue 03 handle configuration changes

olpie101 opened the pull request:
Based on [issue-03](https://github.com/praekelt/mama-android/issues/3)

olpie101 committed:
chore: updated support library dependency

olpie101 committed:
fix: fixed deprecated split function usage

olpie101 committed:
refactor: removed redundant flatmap operator

olpie101 committed:
chore: added support annotations library

olpie101 committed:
refactor: initialize variables in CategoryListAdapter ViewHolder within init

olpie101 committed:
refactor: removed redundant refresh

olpie101 committed:
refactor: hold reference to activity instead of create new instance each time

olpie101 committed:
feat: added empty list state for category

now displays message when no categories are available

olpie101 committed:
feat: continue network call across configuration changes

network calls are not restarted on configuration changes

olpie101 commented:
Manage to maintain connectivity changes across configuration changes. Current solution is a bit of a hack. Will need to come up with something a little more robust.

As result some classes have swapped responsibilities. The Dagger module and component for the main activity is now has the responsibility of fetch or create the necessary observables that must be maintained across state changes.

JayH5 commented on app/build.gradle:
```diff
@@ -8,7 +8,7 @@ buildscript{
         dbFlowVersion = "2.1.0"
         gsonVersion = "2.3"
-        supportLibVersion = "22.2.0"
+        supportLibVersion = "22.2.1"
```
I think we're up to version 23 stable on all these things now

JayH5 commented on app/src/main/kotlin/za/foundation/praekelt/mama/app/App.kt:
```diff
@@ -2,24 +2,60 @@ package za.foundation.praekelt.mama.app
+        bus.register(this)
+    }
+
+    override fun onLowMemory() {
```
This method is not called very reliably in my experience and the way it is used differs across Android versions. Best be sure this is going to work/is necessary.

JayH5 commented on app/src/main/kotlin/za/foundation/praekelt/mama/app/App.kt:
```diff
@@ -2,24 +2,60 @@ package za.foundation.praekelt.mama.app
+        observables.clear()
+    }
+
+    override fun onTerminate() {
```
From the docs:
This method is for use in emulated process environments. It will never be called on a production Android device, where processes are removed by simply killing them; no user code (including this callback) is executed when doing so.

olpie101 replied:
Thanks for the feedback. Realised that I do not need the onLowMem or onTerminate. Observables are relatively small. They contribute to less than 0.1% of the memory usage of memory allocated to my app, which I believe is effectively negligible.

JayH5 commented on app/src/main/kotlin/za/foundation/praekelt/mama/app/activity/MainActivity.kt:
```diff
@@ -60,51 +69,31 @@ public class MainActivity : AppCompatActivity(), AnkoLogger {
         println("end resuming")
     }
 
+    override fun onPause() {
```
Would it be possible to subscribe/unsubscribe in onStart/Stop rather. onResume/Pause gets called a lot...

JayH5 commented on app/src/main/kotlin/za/foundation/praekelt/mama/inject/module/MainActivityModule.kt:
```diff
@@ -66,4 +82,62 @@ class MainActivityModule(val activity: MainActivity) {
+            cached[0] as Observable<Repo>
+        else
+            networkObs.filter { it }
+                    .map { SharedPrefsUtil.getCommitFromSharedPrefs(activity) }
```
This is a lot of steps... maybe a comment or two?

olpie101 committed:
refactor: made RepoPull inherit from Repo

makes rx functions easier to manage

olpie101 committed:
fix: leak caused by holding reference to destroyed activity in clone and update observables

olpie101 committed:
refactor: merged the clone repo observable and update repo observable

noe referred to as repoObs

olpie101 committed:
refactor: refresh category page adapter instead creating a new one when network in MainActivity is complete

applies to whether it is first time and cloning or just an update. This helps maintain scroll position on update

olpie101 committed:
chore: removed block of commented code

olpie101 committed:
refactor: ensure long running observable operations run on the io thread

olpie101 committed:
refactor: remove onTerminate and onLowMemory call in App class

olpie101 committed:
chore: added comments to the MainActivityModule class

olpie101 commented:
Finished, however still has a leak. Learning the hard way that Observables do not mix well with the Activity lifecycle. Need to separate observables from activity using something like a presenter layer.

olpie101 commented:
Merging. Any issues with this branch can be posted here. I will move them to issue https://github.com/praekelt/mama-android/issues/5.
